include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,19 +121,210 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
+                const auto& __group = __self_item.get_group();
+                const auto __item_id = __self_item.get_local_linear_id();
+
+                for (uint16_t __i = 0; __i < __elems_per_item; ++__i)
+                {
+                   auto __global_idx = __i * __wg_size + __item_id;
+                   __lacc[__global_idx] = __global_idx < __n ? __unary_op(__in[__global_idx]) : _ValueType{};
+                }
+
+                __group_scan<_ValueType>(__group, __lacc.get_pointer(), __lacc.get_pointer() + __n, __bin_op, __init);
+
+                for (uint16_t __i = 0; __i < __elems_per_item; ++__i)
+                {
+                   auto __global_idx = __i * __wg_size + __item_id;
+                   if (__global_idx < __n)
+                       __out[__global_idx] = __lacc[__global_idx];
+                }
+            });
+        });
+        __event.wait();
+    }
+
+    template<::uint32_t _ElemsPerItem, ::uint32_t _WGSize, bool _IsFullGroup, typename _Policy, typename _InRng, typename _OutRng, typename _InitType, typename _BinaryOperation, typename _UnaryOp>
+    static void apply(_Policy const & __policy, _InRng __in, _OutRng __out, std::size_t __n, _InitType __init, const _BinaryOperation& __bin_op, const _UnaryOp& __unary_op)
+    {
+        using _RangeValueType = decltype(*__in.begin());
+        using _ValueType = decltype(__bin_op(std::declval<_RangeValueType>(), std::declval<_RangeValueType>()));
+        using _CustomName = typename _Policy::kernel_name;
+        using _GroupScanKernel =
+                     oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__par_backend_hetero::__scan_single_wg_kernel, _CustomName, _BinaryOperation, std::integral_constant<bool, _Inclusive>, std::integral_constant<bool, _IsFullGroup>, _InRng, _OutRng, std::integral_constant<::std::size_t, _ElemsPerItem>, std::integral_constant<std::size_t, _WGSize>>;
+
+
+        constexpr ::uint32_t __elems_per_wg = _ElemsPerItem * _WGSize;
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc = sycl::accessor<_ValueType, 1, sycl::access_mode::read_write, sycl::target::local>(sycl::range<1>{__elems_per_wg}, __hdl);
+
+            __hdl.parallel_for<__par_backend_hetero::__tunable_kernel_name<_GroupScanKernel, __elems_per_wg, _IsFullGroup, _Inclusive>>(sycl::nd_range<1>(_WGSize, _WGSize), [=](sycl::nd_item<1> __self_item) {
+                const auto& __group = __self_item.get_group();
+                const auto& __subgroup = __self_item.get_sub_group();
+                const auto __item_id = __self_item.get_local_linear_id();
+                int __subgroup_id = __subgroup.get_group_id();
+                int __id_in_subgroup = __subgroup.get_local_id();
+                int __subgroup_size = __subgroup.get_local_linear_range();
```

Comment:
Yes, the `int` values here can be `uint16_t`. I should also change the `auto` for the work-item id to `uint16_t` as well, since `get_local_linear_id()` returns a  `size_t`  but this version is only called for sizes up to 2^14.